webpack.config.ts

```diff
@@ -5,92 +5,150 @@ import ESLintPlugin from 'eslint-webpack-plugin'
-const buildDir = path.resolve(rootDir, 'dist')
-const uiDir = path.resolve(srcDir, 'ui')
+const ExtraActionsPlugin = (event: string, callback: (compilation: Compilation) => void) => ({
+  apply: (compiler: Compiler) => {
+    compiler.hooks.afterEmit.tap(event, callback)
+  },
+  userOptions: {},
+  version: 1,
+})
 
-const isDev = process.env.ENVIRONMENT === 'development'
+function createLiveReloadPlugin(): HtmlWebpackPlugin {
+  const server: Server = new Server({ port: RELOAD_PORT })
 
-const pages = fs.readdirSync(uiDir).filter((page) => page !== 'common')
+  return ExtraActionsPlugin('LiveReloadPlugin', () => {
+    server.clients.forEach((client) => {
+      client.send('ping')
+    })
+  })
+}
 
-const pageEntries = pages.reduce((pagesMap, page) => {
-  pagesMap[page] = path.resolve(uiDir, page, 'index.tsx')
+interface WebpackEnvParams {
+  ui: string
+  build: string
+}
 
-  return pagesMap
-}, {})
+module.exports = (env: WebpackEnvParams) => {
+  dotenv.config({ override: true })
 
-const htmlPlugins = pages.map(
-  (page) =>
-    new HtmlWebpackPlugin({
-      filename: `${page}.html`,
-      template: path.resolve(uiDir, page, 'index.html'),
-      chunks: [page],
-    }),
-)
+  const rootDir = path.resolve(__dirname)
+  const srcDir = path.resolve(rootDir, 'src')
+  const buildDir = path.resolve(rootDir, 'dist')
+  const uiDir = path.resolve(srcDir, 'ui')
+  let plugins: HtmlWebpackPlugin[] = []
 
-module.exports = {
-  mode: process.env.ENVIRONMENT,
-  devtool: isDev ? 'inline-source-map' : undefined,
-  resolve: {
-    extensions: ['.tsx', '.ts', '.jsx', '.js', '.json'],
-  },
-  entry: {
-    ...pageEntries,
-    main: path.resolve(srcDir, 'main.ts'),
-    content: path.resolve(srcDir, 'content', 'index.ts'),
-  },
-  output: {
-    filename: '[name].js',
-    publicPath: '/',
-    path: buildDir,
-    clean: true,
-  },
-  module: {
-    rules: [
-      {
-        test: /\.(ts|tsx|js|jsx)$/,
-        exclude: /node_modules/,
-        use: {
-          loader: 'babel-loader',
-          options: {
-            presets: [
-              '@babel/preset-react',
-              '@babel/preset-env',
-              '@babel/typescript',
-              {
-                plugins: ['@babel/plugin-transform-runtime'],
-              },
-            ],
+  const isDev = process.env.ENVIRONMENT === 'development'
+  const isUI = env.ui === 'true'
+  const isBuild = env.build === 'true'
+
+  const pages = fs.readdirSync(uiDir).filter((page) => page !== 'common')
+
+  const pageEntries = pages.reduce((pagesMap, page) => {
+    pagesMap[page] = path.resolve(uiDir, page, 'index.tsx')
+
+    return pagesMap
+  }, {})
+
+  if (isUI || isBuild) {
+    plugins = plugins.concat(
+      pages.map(
+        (page) =>
+          new HtmlWebpackPlugin({
+            filename: `${page}.html`,
+            template: path.resolve(uiDir, page, 'index.html'),
+            chunks: [page],
+          }),
+      ),
+    )
+  }
+
+  if (!isUI || isBuild) {
```

Comment:
why is it not enough to just check `isBuild`? or use `&&` between the flags?

Reply:
Build process is separated into two processes for development. So one `npm start` will watch for service worker scripts and other one for UI pages. That's the only way to avoid refreshing whole extension on all changes, because for UI changes there is no need to reload the extension.

I tried to intercept assets and determine whether reload is needed or not, but there is no way to retrieve asset paths, only file name and source code is available.

So the `isUI` flag indicates that dev process is for UI pages only and `isBuild` is true when process is not executed in 'watch' mode. That means that all files should be compiled in that process.